adapters/repos/db/vector/ssdhelpers/compression.go

```diff
@@ -0,0 +1,448 @@
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/cache"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer"
+	"github.com/weaviate/weaviate/entities/storobj"
+	"github.com/weaviate/weaviate/entities/vectorindex/hnsw"
+)
+
+type CompressorDistancer interface {
+	DistanceToNode(id uint64) (float32, bool, error)
+	DistanceToFloat(vec []float32) (float32, bool, error)
+}
+
+type QuantizerDistancer[T byte | uint64] interface {
+	Distance(x []T) (float32, bool, error)
+	DistanceToFloat(x []float32) (float32, bool, error)
+}
+
+type CompressionDistanceBag interface {
+	Load(ctx context.Context, id uint64) error
+	Distance(x, y uint64) (float32, error)
+}
+
+type QuantizedDistanceBag[T byte | uint64] struct {
+	elements   map[uint64][]T
+	compressor *QuantizedVectorsCompressor[T]
+}
+
+func (bag *QuantizedDistanceBag[T]) Load(ctx context.Context, id uint64) error {
+	var err error
+	bag.elements[id], err = bag.compressor.cache.Get(ctx, id)
+	return err
+}
+
+func (bag *QuantizedDistanceBag[T]) Distance(x, y uint64) (float32, error) {
+	v1, found := bag.elements[x]
+	if !found {
+		return 0, fmt.Errorf("missing id in bag: %d", x)
+	}
+	v2, found := bag.elements[y]
+	if !found {
+		return 0, fmt.Errorf("missing id in bag: %d", y)
+	}
+	return bag.compressor.DistanceBetweenCompressedVectors(v1, v2)
+}
+
+type VectorCompressor interface {
+	Drop() error
+	GrowCache(size uint64)
+	SetCacheMaxSize(size int64)
+	GetCacheMaxSize() int64
+	Delete(ctx context.Context, id uint64)
+	Preload(id uint64, vector []float32)
+	Prefetch(id uint64)
+	PrefillCache()
+
+	PauseCompaction(ctx context.Context) error
+	FlushMemtables(ctx context.Context) error
+	ResumeCompaction(ctx context.Context) error
+	ListFiles(ctx context.Context, basePath string) ([]string, error)
+
+	DistanceBetweenCompressedVectorsFromIDs(ctx context.Context, x, y uint64) (float32, error)
+	DistanceBetweenCompressedAndUncompressedVectorsFromID(ctx context.Context, x uint64, y []float32) (float32, error)
+	NewDistancer(vector []float32) CompressorDistancer
+	NewDistancerFromID(id uint64) CompressorDistancer
+	ReturnDistancer(distancer CompressorDistancer)
+	NewBag() CompressionDistanceBag
+
+	ExposeFields() PQData
+}
+
+type quantizer[T byte | uint64] interface {
+	DistanceBetweenCompressedVectors(x, y []T) (float32, error)
+	DistanceBetweenCompressedAndUncompressedVectors(x []float32, encoded []T) (float32, error)
+	Encode(vec []float32) []T
+	NewQuantizerDistancer(a []float32) QuantizerDistancer[T]
+	NewCompressedQuantizerDistancer(a []T) QuantizerDistancer[T]
+	ReturnQuantizerDistancer(distancer QuantizerDistancer[T])
+	CompressedBytes(compressed []T) []byte
+	FromCompressedBytes(compressed []byte) []T
+	ExposeFields() PQData
+}
+
+type QuantizedVectorsCompressor[T byte | uint64] struct {
+	cache           cache.Cache[T]
+	compressedStore *lsmkv.Store
+	quantizer       quantizer[T]
+}
+
+func (compressor *QuantizedVectorsCompressor[T]) Drop() error {
```

Comment:
I see Drop being used in hnsw::Drop and hnsw::Shutdown. If vector index is dropped I would assume related bucket is removed as well (though dropping bucket does not exist yet).
Bucket::Shutdown does not remove bucket's files.